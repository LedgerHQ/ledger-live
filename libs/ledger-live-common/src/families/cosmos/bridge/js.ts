import createTransaction from "../js-createTransaction";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import getTransactionStatus from "../js-getTransactionStatus";
import prepareTransaction from "../js-prepareTransaction";
import signOperation from "../js-signOperation";
import { sync, scanAccounts } from "../js-synchronisation";
import updateTransaction from "../js-updateTransaction";
import type { CosmosValidatorItem, Transaction } from "../types";
import cosmosValidatorsManager from "../validators";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { defaultCosmosAPI } from "../api/Cosmos";
import {
  asSafeCosmosPreloadData,
  setCosmosPreloadData,
} from "../preloadedData";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const getPreloadStrategy = (_currency) => ({
  preloadMaxAge: 30 * 1000,
});

const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload: async () => {
    const validators = await cosmosValidatorsManager.getValidators();
    setCosmosPreloadData({
      validators,
    });
    return Promise.resolve({
      validators,
    });
  },
  hydrate: (data: { validators?: CosmosValidatorItem[] }) => {
    if (!data || typeof data !== "object") return;
    const { validators } = data;
    if (
      !validators ||
      typeof validators !== "object" ||
      !Array.isArray(validators)
    )
      return;
    cosmosValidatorsManager.hydrateValidators(validators);
    setCosmosPreloadData(asSafeCosmosPreloadData(data));
  },
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  estimateMaxSpendable,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast: defaultCosmosAPI.broadcast,
};

export default {
  currencyBridge,
  accountBridge,
};
