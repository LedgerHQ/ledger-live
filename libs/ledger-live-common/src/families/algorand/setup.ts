// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
