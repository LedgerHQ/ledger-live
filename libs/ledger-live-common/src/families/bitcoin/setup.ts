// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { firstValueFrom, from } from "rxjs";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import Btc from "@ledgerhq/hw-app-btc";
import { createBridges } from "@ledgerhq/coin-bitcoin/bridge/js";
import type { SignerContext } from "@ledgerhq/coin-bitcoin/signer";
import makeCliTools from "@ledgerhq/coin-bitcoin/cli-transaction";
import bitcoinResolver from "@ledgerhq/coin-bitcoin/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-bitcoin/hw-signMessage";
import { Transaction } from "@ledgerhq/coin-bitcoin/types";
import { GetAddressOptions, Resolver } from "../../hw/getAddress/types";
import { withDevice } from "../../hw/deviceAccess";
import { startSpan } from "../../performance";
import { GetAddressFn } from "@ledgerhq/coin-framework/bridge/getAddressWrapper";
import { getCurrencyConfiguration } from "../../config";
import { BitcoinConfigInfo } from "@ledgerhq/coin-bitcoin/lib/config";
import { SignMessage } from "../../hw/signMessage/types";

const createSigner = (transport: Transport, currency: CryptoCurrency) => {
  return new Btc({ transport, currency: currency.id });
};

const signerContext: SignerContext = <T>(
  deviceId: string,
  crypto: CryptoCurrency,
  fn: (signer: Btc) => Promise<T>,
): Promise<T> =>
  firstValueFrom(
    withDevice(deviceId)((transport: Transport) => from(fn(createSigner(transport, crypto)))),
  );

const getCurrencyConfig = (currency: CryptoCurrency) => {
  return { info: getCurrencyConfiguration<BitcoinConfigInfo>(currency) };
};

const perfLogger = {
  startSpan,
};

const bridge: Bridge<Transaction> = createBridges(signerContext, perfLogger, getCurrencyConfig);

export function createMessageSigner(): SignMessage {
  return (transport, account, messageData) => {
    const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
    return signMessage(signerContext)("", account, messageData);
  };
}

const messageSigner = {
  signMessage: createMessageSigner(),
};

const resolver: Resolver = (
  transport: Transport,
  addressOpt: GetAddressOptions,
): ReturnType<GetAddressFn> => {
  const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
  return bitcoinResolver(signerContext)("", addressOpt);
};

const cliTools = makeCliTools();

export { bridge, cliTools, resolver, messageSigner, signerContext };
