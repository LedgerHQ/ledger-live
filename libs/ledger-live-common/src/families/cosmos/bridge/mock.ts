import { BigNumber } from "bignumber.js";
import {
  NotEnoughBalance,
  RecipientRequired,
  InvalidAddress,
  FeeTooHigh,
} from "@ledgerhq/errors";
import type {
  CosmosAccount,
  CosmosValidatorItem,
  StatusErrorMap,
  Transaction,
} from "../types";
import {
  scanAccounts,
  signOperation,
  broadcast,
  sync,
  isInvalidRecipient,
} from "../../../bridge/mockHelpers";
import {
  setCosmosPreloadData,
  asSafeCosmosPreloadData,
} from "../preloadedData";
import { getMainAccount } from "../../../account";
import mockPreloadedData from "../preloadedData.mock";
import { makeAccountBridgeReceive } from "../../../bridge/mockHelpers";
import type {
  Account,
  AccountBridge,
  CurrencyBridge,
} from "@ledgerhq/types-live";
const receive = makeAccountBridgeReceive();

const defaultGetFees = (a, t) =>
  (t.fees || new BigNumber(0)).times(t.gas || new BigNumber(0));

const createTransaction = (): Transaction => ({
  family: "cosmos",
  mode: "send",
  amount: new BigNumber(0),
  recipient: "",
  fees: null,
  gas: null,
  memo: null,
  validators: [],
  sourceValidator: null,
  networkInfo: null,
  useAllAmount: false,
});

const updateTransaction = (t, patch) => ({ ...t, ...patch });

const estimateMaxSpendable = ({ account, parentAccount, transaction }) => {
  const mainAccount = getMainAccount(account, parentAccount);
  const estimatedFees = transaction
    ? defaultGetFees(mainAccount, transaction)
    : new BigNumber(5000);
  return Promise.resolve(
    BigNumber.max(0, account.balance.minus(estimatedFees))
  );
};

const getTransactionStatus = (account: Account, t: Transaction) => {
  const errors: StatusErrorMap = {};
  const warnings: StatusErrorMap = {};
  const useAllAmount = !!t.useAllAmount;
  const estimatedFees = defaultGetFees(account, t);
  const totalSpent = useAllAmount
    ? account.balance
    : new BigNumber(t.amount).plus(estimatedFees);
  const amount = useAllAmount
    ? account.balance.minus(estimatedFees)
    : new BigNumber(t.amount);

  if (amount.gt(0) && estimatedFees.times(10).gt(amount)) {
    warnings.feeTooHigh = new FeeTooHigh();
  }

  // Fill up transaction errors...
  if (totalSpent.gt(account.balance)) {
    errors.amount = new NotEnoughBalance();
  }

  // Fill up recipient errors...
  if (!t.recipient) {
    errors.recipient = new RecipientRequired("");
  } else if (isInvalidRecipient(t.recipient)) {
    errors.recipient = new InvalidAddress("");
  }

  return Promise.resolve({
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  });
};

const prepareTransaction = async (
  a: CosmosAccount,
  t: Transaction
): Promise<Transaction> => {
  if (!t.networkInfo) {
    return {
      ...t,
      gas: new BigNumber(1),
      fees: new BigNumber(500),
      networkInfo: {
        family: "cosmos",
        fees: new BigNumber(500),
      },
    };
  }

  return t;
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: () => {
    setCosmosPreloadData(mockPreloadedData);
    return Promise.resolve(mockPreloadedData);
  },
  hydrate: (data: { validators?: CosmosValidatorItem[] }) => {
    setCosmosPreloadData(asSafeCosmosPreloadData(data));
  },
};
export default {
  currencyBridge,
  accountBridge,
};
