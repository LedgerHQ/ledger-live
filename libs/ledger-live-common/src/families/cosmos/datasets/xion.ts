import { fromTransactionRaw } from "@ledgerhq/coin-cosmos/transaction";
import { CurrenciesData } from "@ledgerhq/types-live";
import { BigNumber } from "bignumber.js";
import type { Transaction } from "../types";

const dataset: CurrenciesData<Transaction> = {
  FIXME_ignoreAccountFields: ["cosmosResources", "operationsCount", "operations"],
  FIXME_ignorePreloadFields: ["validators"],
  scanAccounts: [
    {
      name: "xion seed 1",
      apdus: `
      => 55040000190478696f6e2c00008076000080000000800000000000000000
      <= 03d5e0ebb3f1ae2afe87e5d5a24b5029a59cc12f8fd1056840091b2f0b97e54e8378696f6e3167796175766c3434713261706e33753361756a6d333671387a726a373476727968337934657a9000
      => 55040000190478696f6e2c00008076000080000000800000000000000000
      <= 03d5e0ebb3f1ae2afe87e5d5a24b5029a59cc12f8fd1056840091b2f0b97e54e8378696f6e3167796175766c3434713261706e33753361756a6d333671387a726a373476727968337934657a9000
      => 55040000190478696f6e2c00008076000080010000800000000000000000
      <= 02c8f29e73a2c99d11de0544615a87afc6c6575d22746af18a8ef733066f088b1f78696f6e3176326d70306d376b3936646d3971763630666b737063716c7a706b7a72776e657176663036399000
      `,
    },
  ],
  accounts: [
    {
      FIXME_tests: ["balance is sum of ops", "pendingOperations are cleaned up"],
      raw: {
        id: "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
        seedIdentifier: "03d5e0ebb3f1ae2afe87e5d5a24b5029a59cc12f8fd1056840091b2f0b97e54e83",
        xpub: "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
        freshAddress: "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
        freshAddressPath: "44'/118'/0'/0/0",
        name: "Xion 1",
        derivationMode: "",
        index: 0,
        currencyId: "xion",
        operationsCount: 2,
        operations: [],
        pendingOperations: [],
        lastSyncDate: "",
        balance: "676335",
        spendableBalance: "562180",
        blockHeight: 4486981,
      },
      transactions: [
        {
          name: "Normal transaction",
          transaction: fromTransactionRaw({
            amount: "10000",
            recipient: "xion108uy5q9jt59gwugq5yrdhkzcd9jryslmr32fql",
            useAllAmount: false,
            family: "cosmos",
            mode: "send",
            networkInfo: null,
            fees: "8744",
            gas: "87440",
            validators: [],
            memo: "",
            sourceValidator: "",
          }),
          expectedStatus: () => ({
            errors: {},
            warnings: {},
            amount: BigNumber("10000"),
          }),
        },
      ],
    },
  ],
};

export default dataset;
