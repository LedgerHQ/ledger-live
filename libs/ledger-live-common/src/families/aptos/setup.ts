// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
