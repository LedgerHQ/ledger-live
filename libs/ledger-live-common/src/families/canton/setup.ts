// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-canton/bridge/index";
import { getEnv } from "@ledgerhq/live-env";
import Transport from "@ledgerhq/hw-transport";
import { CantonSigner } from "@ledgerhq/coin-canton";
import cantonResolver from "@ledgerhq/coin-canton/signer";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-canton/test/cli";
import { CantonCoinConfig } from "@ledgerhq/coin-canton/config";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-canton/types";
import { LegacySignerCanton } from "@ledgerhq/live-signer-canton";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import cantonBridgeMock from "./bridge/mock";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "../../currencies";

const createSigner: CreateSigner<CantonSigner> = (transport: Transport) => {
  return new LegacySignerCanton(transport);
};

const getCurrencyConfig = () => {
  // Use devnet for development/testing
  const currencyId = getEnv("MOCK") ? "canton_network_devnet" : "canton_network_devnet";
  return getCurrencyConfiguration<CantonCoinConfig>(getCryptoCurrencyById(currencyId));
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = getEnv("MOCK")
  ? cantonBridgeMock
  : createBridges(executeWithSigner(createSigner), getCurrencyConfig);

const resolver: Resolver = createResolver(createSigner, cantonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
