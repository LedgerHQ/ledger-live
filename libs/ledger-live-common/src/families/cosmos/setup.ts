// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-cosmos/bridge/index";
import makeCliTools from "@ledgerhq/coin-cosmos/cli";
import { CosmosCoinConfig } from "@ledgerhq/coin-cosmos/config";
import cosmosResolver from "@ledgerhq/coin-cosmos/hw-getAddress";
import { CosmosAccount, Transaction, TransactionStatus } from "@ledgerhq/coin-cosmos/types/index";
import { CosmosSigner } from "@ledgerhq/coin-cosmos/types/signer";
import Cosmos from "@ledgerhq/hw-app-cosmos";
import Transport from "@ledgerhq/hw-transport";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import type { Bridge } from "@ledgerhq/types-live";
import { CosmosApp } from "@zondax/ledger-cosmos-js";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { getCurrencyConfiguration } from "../../config";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<CosmosSigner> = (transport: Transport) => {
  const cosmos = new CosmosApp(transport);
  const hwCosmos = new Cosmos(transport);
  return {
    getAddressAndPubKey: cosmos.getAddressAndPubKey.bind(cosmos),
    sign: cosmos.sign.bind(cosmos),
    getAddress: hwCosmos.getAddress.bind(hwCosmos),
  };
};

const getCurrencyConfig = (currency?: CryptoCurrency) => {
  if (!currency) {
    throw new Error("No currency provided");
  }
  return getCurrencyConfiguration<CosmosCoinConfig>(currency);
};
const bridge: Bridge<Transaction, CosmosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCurrencyConfig,
);

const resolver: Resolver = createResolver(createSigner, cosmosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
