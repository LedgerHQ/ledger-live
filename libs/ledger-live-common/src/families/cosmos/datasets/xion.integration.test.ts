import { DatasetTest } from "@ledgerhq/types-live";
import { testBridge } from "../../../__tests__/test-helpers/bridge";
import "../../../__tests__/test-helpers/setup";
import type { Transaction } from "../types";
import xion from "./xion";

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    xion,
  },
};

testBridge(dataset);
