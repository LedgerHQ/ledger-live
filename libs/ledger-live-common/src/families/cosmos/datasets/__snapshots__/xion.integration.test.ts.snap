// Jest Snapshot v1, https://goo.gl/fbAQLP
exports[`xion currency bridge scanAccounts xion seed 1 1`] = `
[
  {
    "balance": "411630",
    "currencyId": "xion",
    "derivationMode": "",
    "freshAddress": "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
    "freshAddressPath": "44'/118'/0'/0/0",
    "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
    "index": 0,
    "pendingOperations": [],
    "seedIdentifier": "03d5e0ebb3f1ae2afe87e5d5a24b5029a59cc12f8fd1056840091b2f0b97e54e83",
    "spendableBalance": "0",
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
    "xpub": "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
  },
  {
    "balance": "0",
    "currencyId": "xion",
    "derivationMode": "",
    "freshAddress": "xion1v2mp0m7k96dm9qv60fkspcqlzpkzrwneqvf069",
    "freshAddressPath": "44'/118'/1'/0/0",
    "id": "js:2:xion:xion1v2mp0m7k96dm9qv60fkspcqlzpkzrwneqvf069:",
    "index": 1,
    "pendingOperations": [],
    "seedIdentifier": "03d5e0ebb3f1ae2afe87e5d5a24b5029a59cc12f8fd1056840091b2f0b97e54e83",
    "spendableBalance": "0",
    "swapHistory": [],
    "syncHash": undefined,
    "used": false,
    "xpub": "xion1v2mp0m7k96dm9qv60fkspcqlzpkzrwneqvf069",
  },
]
`;
exports[`xion currency bridge scanAccounts xion seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4488607,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper1t8ysd9cg6s38ak9xfadkjtvfv9uczvrn8wzv77",
            "amount": "209175",
          },
        ],
      },
      "fee": "8410",
      "hasFailed": false,
      "hash": "0DFDF9C21EE698B9918AA30CF4B49000028943254A3FFAF36DDB0E053687CD9A",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-0DFDF9C21EE698B9918AA30CF4B49000028943254A3FFAF36DDB0E053687CD9A-DELEGATE",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 6,
      "type": "DELEGATE",
      "value": "8410",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4488564,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper1t8ysd9cg6s38ak9xfadkjtvfv9uczvrn8wzv77",
            "amount": "66225",
          },
        ],
      },
      "fee": "7206",
      "hasFailed": false,
      "hash": "2A2B7691BAD01B57C7778300C4067B0832FE4AE34D256A65F34168C7CA891D15",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-2A2B7691BAD01B57C7778300C4067B0832FE4AE34D256A65F34168C7CA891D15-DELEGATE",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 4,
      "type": "DELEGATE",
      "value": "7206",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4486960,
      "extra": {
        "memo": "Ledger Live",
      },
      "fee": "4349",
      "hasFailed": false,
      "hash": "3E0ED76323E0606A7AFABC7B1C7DDA4D07FE78F6B4CFCFF141F37F4532CF3267",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-3E0ED76323E0606A7AFABC7B1C7DDA4D07FE78F6B4CFCFF141F37F4532CF3267-IN",
      "recipients": [
        "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
      ],
      "senders": [
        "xion1g84934jpu3v5de5yqukkkhxmcvsw3u2as0krm5",
      ],
      "transactionSequenceNumber": 3,
      "type": "IN",
      "value": "684211",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4488530,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper1t8ysd9cg6s38ak9xfadkjtvfv9uczvrn8wzv77",
            "amount": "10",
          },
        ],
      },
      "fee": "6371",
      "hasFailed": false,
      "hash": "62DE7FF0CFFFE7FA3AFBC5E4B31A234494B249C2F6B61AB72094E336BE93A7F3",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-62DE7FF0CFFFE7FA3AFBC5E4B31A234494B249C2F6B61AB72094E336BE93A7F3-REWARD",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 2,
      "type": "REWARD",
      "value": "10",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4488578,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper18hj7x8dlchh77dd2k9wr7dqvygmmpxrpxtzql9",
            "amount": "22075",
          },
        ],
      },
      "fee": "7219",
      "hasFailed": false,
      "hash": "7E470E2AB00F4BC595AD9F9416C4ECCF6C987818AB6A8F5A94EAC04241D5A4E3",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-7E470E2AB00F4BC595AD9F9416C4ECCF6C987818AB6A8F5A94EAC04241D5A4E3-DELEGATE",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 5,
      "type": "DELEGATE",
      "value": "7219",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4488540,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper1t8ysd9cg6s38ak9xfadkjtvfv9uczvrn8wzv77",
            "amount": "114155",
          },
        ],
      },
      "fee": "9314",
      "hasFailed": false,
      "hash": "82E94B0710F4EDC16D269560263E0185F1FB877A94F857D37615BD0814ACD335",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-82E94B0710F4EDC16D269560263E0185F1FB877A94F857D37615BD0814ACD335-UNDELEGATE",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 3,
      "type": "UNDELEGATE",
      "value": "9314",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4487736,
      "extra": {
        "memo": "Ledger Live",
      },
      "fee": "4466",
      "hasFailed": false,
      "hash": "8DA137619218388751C60C5515062F8F313143AF493C7CE3D839C210A94A6E46",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-8DA137619218388751C60C5515062F8F313143AF493C7CE3D839C210A94A6E46-OUT",
      "recipients": [
        "xion1rs97j43nfyvc689y5rjvnnhrq3tes6ghe4j84x",
      ],
      "senders": [
        "xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez",
      ],
      "transactionSequenceNumber": 1,
      "type": "OUT",
      "value": "226195",
    },
    {
      "accountId": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:",
      "blockHash": null,
      "blockHeight": 4486979,
      "extra": {
        "memo": "Ledger Live",
        "validators": [
          {
            "address": "xionvaloper1t8ysd9cg6s38ak9xfadkjtvfv9uczvrn8wzv77",
            "amount": "114155",
          },
        ],
      },
      "fee": "7876",
      "hasFailed": false,
      "hash": "C0042A42F14C0454C8F5D46BA16C50B6810E5E3504E25C731BB03DD688083B47",
      "id": "js:2:xion:xion1gyauvl44q2apn3u3aujm36q8zrj74vryh3y4ez:-C0042A42F14C0454C8F5D46BA16C50B6810E5E3504E25C731BB03DD688083B47-DELEGATE",
      "recipients": [],
      "senders": [],
      "transactionSequenceNumber": 0,
      "type": "DELEGATE",
      "value": "7876",
    },
  ],
  [],
]
`;